src/main/io/dataEdition.h

```diff
@@ -0,0 +1,220 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * Authors:
+ * Dominic Clifton/Hydra
+ * Carsten Giesen
+ * Adam Majerczyk (majerczyk.adam@gmail.com)
+ * Texmode add-on by Michi (mamaretti32@gmail.com)
+ *
+ *  Created on: december 2015
+ *  Author: Pierre Archambault
+ */
+
+#define NO_COMMAND			0x00
+#define EXIT_EDITION_MODE	0x01	// THR_LO + YAW_CE + PIT_LO + ROL_LO
+#define ENTER_EDITION_MODE	0x02	// THR_LO + YAW_CE + PIT_LO + ROL_HI
+#define INC_COMMAND    		0x0B	// THR_LO + YAW_CE + PIT_HI + ROL_CE
+#define DEC_COMMAND			0x0D	// THR_LO + YAW_CE + PIT_LO + ROL_CE
+#define SET_COMMAND			0x09	// THR_LO + YAW_HI + PIT_HI + ROL_CE
+#define NEXT_COMMAND		0x0E	// THR_LO + YAW_CE + PIT_CE + ROL_HI
+#define PREV_COMMAND		0x07	// THR_LO + YAW_CE + PIT_CE + ROL_LO
+
+#define MILLISECONDS_IN_A_SECOND 	1000
+
+#define DEFAULT_DISPLAY_COLUMNS 	21
+#define DEFAULT_DISPLAY_LINES 		8
+
+
+typedef enum {
+    PIXEL_NORMAL = 0,
+    PIXEL_INVERTED = (1 << 0),
+    PIXEL_MIXED = (1 << 1)
+} pixelColor_e;
+
+#define LOWER_FLASH_SIZE_FOR_FULL_EDITION_MENUS    128
+#define DATA_EDITION_FULL_VERSION    (defined(FLASH_SIZE) && (FLASH_SIZE > LOWER_FLASH_SIZE_FOR_FULL_EDITION_MENUS))
+
+#if (!DATA_EDITION_FULL_VERSION)
+static const char * const pageNames[] = {
```

Comment:
well..
I moved all these definitions (except codes of stick combinations needed 
in rc_controls)  and declarations of tables to the .c file and the 
resulting final code size is unchanged at all..


Le 13/02/2016 12:52, Petr Ledvina a écrit :
>
> In src/main/io/dataEdition.h 
> <https://github.com/cleanflight/cleanflight/pull/1654#discussion_r52826968>:
>
> > +
> > +#define DEFAULT_DISPLAY_COLUMNS 	21
> > +#define DEFAULT_DISPLAY_LINES 		8
> > +
> > +
> > +typedef enum {
> > +    PIXEL_NORMAL = 0,
> > +    PIXEL_INVERTED = (1 << 0),
> > +    PIXEL_MIXED = (1 << 1)
> > +} pixelColor_e;
> > +
> > +#define LOWER_FLASH_SIZE_FOR_FULL_EDITION_MENUS    128
> > +#define DATA_EDITION_FULL_VERSION    (defined(FLASH_SIZE) && (FLASH_SIZE > LOWER_FLASH_SIZE_FOR_FULL_EDITION_MENUS))
> > +
> > +#if (!DATA_EDITION_FULL_VERSION)
> > +static const char * const pageNames[] = {
>
> Compiling is fine - each .c file gets it's own copy of array with 
> pointer to strings.
> String themselves are allocated only once (C linker merges identical 
> strings), but the array can take considerable space.
> It can be 20 bytes per string (the .h file is included in five .c files).
> And this if for each static const char * in this file - each string in 
> this file ..
>
> Try adding |static const char * const dummy[] = { "CH1" };| into this 
> file and compare resulting file size ..
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cleanflight/cleanflight/pull/1654/files#r52826968>.
>